feat(Form): base implementation

mnajdova opened the pull request:
This PR aims to create the base implementation for the Form component. Here is a short RFC for the proposed API.

# Form

This component will be used for displaying a set of related user input fields in a structured way. In the implementation the Form component will use the Grid component for customizing the appearance of the fields inside.

### TODO

- [x] Conformance test
- [ ] Minimal doc site example
- [ ] Stardust base theme
- [ ] Teams Light theme
- [ ] Teams Dark theme
- [ ] Teams Contrast theme
- [ ] Confirm RTL usage
- [ ] [W3 accessibility](https://www.w3.org/standards/webdesign/accessibility) check
- [ ] [Stardust accessibility](https://github.com/stardust-ui/accessibility) check
- [ ] Update glossary props table
- [ ] Update the CHANGELOG.md

# API Proposal

## as: defaultValue: 'form'

An element type to render as (string or function).

## className

Additional classes.

## action

The HTML form action.

## onSubmit

The HTML form submit handler. 

## fields

Array shorthand for the fields inside the form.

## message: Segment shorthand prop

Will show a message inside the form (error, success, warning etc.)
**UPDATE**: the message prop will be implemented to the Form.Field component instead of here.

## size:  enum: sizes

Size of the text and elements inside the form. Should be consistent with the Text's sizes.
**UPDATE**: this will be addressed in a consequent PR, as we still haven't decided on the way it should be implemented.

## loading: PropTypes.bool

Automatically show a loading indicator.
**UPDATE**: after discussion, decided that this prop will not be introduced. See comment: https://github.com/stardust-ui/react/pull/353#discussion_r225130590

## styles: ComponentPartStyle

## variables?: ComponentVariablesInput

# Form.Field

A field is a form element containing a label and an input.

## as

An element type to render as (string or function).

## className

Additional classes.

## content: customPropTypes.contentShorthand
Shorthand for primary content.

## control: customPropTypes.some([PropTypes.func, PropTypes.oneOf(['button', 'input', 'select', 'textarea'])])
A form control component (i.e. Input, RadioGroup) or HTML tagName (i.e. 'input').
Extra FormField props are passed to the control component.
Mutually exclusive with children.

## error: PropTypes.bool

Individual fields may display an error state.
UPDATE: this prop will be added in a consequent PR that will address the validation aspects. 

## inline: PropTypes.bool

A field can have its label next to instead of above it.

## label: Text with default as='label' shorthand

Mutually exclusive with children.

## required: PropTypes.bool

A field can show that input is mandatory.

## styles: ComponentPartStyle

## variables?: ComponentVariablesInput

By default the Form field will render the following HTML structure giving label and a input control:

```html
<div> <label for='inputID'>Label text </label><input id='inputID'.../> </div>
```

For the components that doesn't need label (like the Button), or are providing the label together with the component the label will be rendered according to that component implementation.

Commit:
-init branch

Commit:
-initial implementation of Form and FormField

Commit:
-small fix

Commit:
-added required prop in the form field
-refactored the styling for the inline prop

Commit:
-added size context which is used in the FormField for setting the size prop of the label

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,147 @@
+    // No Control
+    // ----------------------------------------
+
+    if (_.isNil(control)) {
```
nit: seems that it could be simplified to `if (control) ... ` without forfeiting correctness - to make this check be consistent with other ones used in this file

kuzhelov commented on docs/src/examples/components/Form/Types/FormExample.tsx:
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
+import { Form, Input, Button } from '@stardust-ui/react'
+
+const FormExample = () => (
```
we should consider to support Shorthand API as well - probably, something that you've already thought of, mentioning just for the sake of align our thoughts on that :) In that case Form will provide an `items` prop (or `fields` one) that will accept an array of `FormField` shorthands.

mnajdova replied:
Already talk with Hugh on this. This was my initial idea (having fields array for defining everything inside the Form), but he had a good argument: 'Form is effectively just wrapper around some controls - so what's inside is pretty free.' I am okay with adding this and we can see further if there will be any need for this prop.

mnajdova replied:
Added fields prop in the form.

kuzhelov replied:
yes, but there is a counterargument to this point - we could add _anything_ as Form.Field, as long as it is a shorthand. So, without limiting the cases covered, we would introduce the `fields/items` API that would help user in vast majority of cases. 

Another thing that goes in favour of that - consider what we do have for `Chat` now, with `ChatItem` component - essentially, the story is exactly the same, as each `ChatItem` item could be heavily customized and, thus, doesn't limit ability to generalize for the component :)

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,133 @@
+    className: PropTypes.string,
+
+    /** The columns of the grid with a space-separated list of values. The values represent the track size, and the space between them represents the grid line. */
+    columns: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
I would rather consider to remove layout-related props from the Form's API for now, as am pretty sure that we would be able to address quite wide set of scenarios by
- introducing base 'vertical form' layout support as a default one
- ability to customize form styles for the client - not that with that it would be possible for the client to provide custom layout styles for each Form.Field element
- potentially add ability to introduce support for horizontal forms

Not saying that **_for sure we won't need Grid support for form's layout_** - the only thing that stating is that _for now_ we, probably, don't need it

mnajdova replied:
This seemed like the easier (for implementation and usage) layouting mechanism support out of the box for the form component. With this by default we are supporting the vertical, multi-column and horizontal layout. We might introduce this as a consequent PR too, after the initial implementation is merged. What are your thoughts on this? And, yes, I would definitelly use the Grid in the background anyway for aligning the elements

kuzhelov replied:
yes - my point is rather about how this layou mechanism would be exposed. And, given that this is a question that should involve some consideration phases, I would not introduce any layout aspects to the API in this PR

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,133 @@
+    onSubmit: PropTypes.func,
+
+    /** Automatically show a loading indicator. */
+    loading: PropTypes.bool,
```
emm, not sure that it should be part of the default form interface. To me `loading` seems to be something that introduces behavioural aspects to the form - and, thus, see this `loading` aspect to be used as rather _decorator_ for the general `Form` component.

mnajdova replied:
Make sense, will remove this prop for now. It seems like a good concept, but in the end introducing this here would then bring out question as why only this component has the loading indicator, so for this moment I agree that it should not be part of the Form component.

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,133 @@
+    action: PropTypes.string,
+
+    /** The size for the Text component */
+    size: PropTypes.oneOf(['smallest', 'smaller', 'small', 'medium', 'large', 'larger', 'largest']),
```
not sure that it should be exposed for `Form` component, especially given that this prop, in fact, defines `Text` size (we should ensure prop types consistency here, by the way then). Pretty sure that it was added for reason - could you, please, suggest which case you've tried to address with that?

mnajdova replied:
The idea for adding this prop was providing out of the box sizing mechanism of all items inside the form. For example, if the user wants the form to have bigger size then regular, applying only one prop on the form, should scale all elements inside (for now it is just the Text components - which are used for the labels, but we should further adjust the sizes of the input and other elements which are going to be used as FormFields) If this property is not added here, then the user should be himself responsible for correct sizing of all elements inside the Form. Please see the latest commit on the PR, which is introducing the SizeContext that is used inside the FormField component.

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,147 @@
+  className?: string
+  label?: ShorthandValue
+  renderLabel?: ShorthandRenderFunction
+  control?: React.ReactType<any>
```
yes, agree, this should considered to be a shorthand.

mnajdova replied:
The logic for generating the control is a bit tricky. The control can be some component (we still don't have all components that can appear in the form implemented) or some HTML element (checkbox, button etc.). This is why this is just a type, and for consistency all properties available in the FormField are applied to the control element. What would be your proposal for using the shorthand in this case? Would be use the shorthand as one specific component and then all other components will always send the <Component ...props /> or some function and be responsible for applying all props inside that component?

kuzhelov replied:
Here is the thing that I am concerned about - how we would be able to customize styles of this component? Yes, while it would be possible (with current approach) to pass `styles` to `Form.Field`, and those, if I've understood you correctly, will be passed to `content` element - for regular HTML element, however, we will have a problem that we would provide unknown `styles` attribute. This is the reason why I see this boundary should be clearly advertised - specifically, by providing the following semantics to the client: here is the `control` slot, please, provide all the necessary component props and style customizations there.

So, I would suggest to use `createHTMLInput` method from `factories.tsx` as a value for `content` - and even given that we have some problems with its implementation now, with this being fixed we should end up with correct implementation for `Form.Field` as well

mnajdova replied:
The `control` is converted to shorthand, and `controlType` prop is added that is the previous `control` prop. Depending on the `controlType`, different slot shorthand factories are created. @Bugaa92 I need your input whether this is the right approach. And yeah, I will need to add some custom tests for this prop, after we decide on the API used.

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,147 @@
+} from '../../../types/utils'
+import Text from '../Text'
+
+export interface IFormFieldProps {
```
one prop that I see missing here is, probably, `errorMessage` one (or general `message`) - the text that will be displayed if field value's validation had failed/passed (text below the input field on the picture): 

![image](https://user-images.githubusercontent.com/9024564/46948826-9ee48300-d07f-11e8-857b-2a1f7a151d86.png)

![image](https://user-images.githubusercontent.com/9024564/46949190-e28bbc80-d080-11e8-8c30-13c84a5aea5f.png)


Might expect that it will be part of the 'extended' `Input` component, but not completely sure, as validation stuff seems to be something that is rather form-related. In either case, do we have any plans to support it?

mnajdova replied:
I introduced the message prop on the form itself, but you are right, it makes more sense to have this inside the field, as different fields might introduce different text for each of them. I will add this as a plain text shorthand that will be displayed below the input for now, and we can iterate on it later, once the validation aspects will be added.

mnajdova replied:
Added message prop for the form field component that appears below the input.

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,136 @@
+  }): React.ReactNode {
+    const { as, rows, columns, action, children, content, size } = this.props
+    return (
+      <SizeContext.Provider value={{ size }}>
```
would rather try to avoid this in general, and certainly at the early stages. Could we list the cases where we need this sizing, and after that think about whether there are alternative ways to achieve the same goal?

mnajdova replied:
This was the reason why I added the sizing mechanism: https://github.com/stardust-ui/react/pull/353#discussion_r225133012 I agree that we may want to address this in some consequent PR, and not the initial implementation.

mnajdova replied:
Reverted the changes for the sizing mechanism from this PR. We can come back to these in some consequent PR.

Commit:
-small fixes

Commit:
-added fields as collection shorthand prop in the form component
-removed sizing logic

Commit:
-renamed renderItem to renderField

Commit:
-added message prop to the form field component

Commit:
-added content prop to the Form and FormField components

Commit:
Merge branch 'master' into feat/form-base

# Conflicts:
#	CHANGELOG.md

Commit:
-remove content from the FormField and used label in it

Commit:
-removed Grid usage from the Form
-improved styles for the FormField

Commit:
-added control type to the form field and changed control prop to be shorthand

Commit:
-added some tests

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+     */
+    controlType: customPropTypes.some([
+      PropTypes.func,
+      PropTypes.oneOf(['button', 'input', 'select', 'textarea']),
```
can we use these to define the `controlType ` prop in the interface?

mnajdova replied:
We don't know all components that can be defined there upfront, that is why the type is `React.ReactType<any>`. We know just for the primitive types that can be one of those.

bmdalex replied:
nit: here's what I meant: since
```tsx
type ReactType<P = any> = string | ComponentType<P>
```
our type could be more precise:
```tsx
controlType: ComponentType<any> | 'button' | 'input' | 'select' | 'textarea'
```
so that the user gets better suggestions; still nit, no need to address

bmdalex replied:
NA anymore

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+      controlType,
+    } = this.props
+
+    let labelContent = label
```
```suggestion
    const labelContent = required
      ? typeof label === 'string'
        ? `${label} *`
        : typeof label === 'object' && (label as any).content
          && { ...label, content: `${(label as any).content} *` }
      : label
```

mnajdova replied:
Will refactor this using two different elements

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+      render: renderMessage,
+    })
+
+    const factoryMethod = createSlotFactory(controlType, name => ({ name }))
```
I think this is ok for now, but we might move this `createSlotFactory` to `lib/factories.tsx` or something similar; I understand the usage now

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+    // Check box or radio (label should appear after the input)
+    // --------------------------------------------------------
+
+    if (controlType && (type === 'checkbox' || type === 'radio')) {
```
`controlType` check not needed
```suggestion
    if (type === 'checkbox' || type === 'radio') {
```

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+        return (
+          <ElementType {...rest} className={classes.root}>
+            {childrenExist(children) && children}
+            {messageElement}
```
I thought we're not supposed to change the `children` if they are provided; with this condition, `{messageElement}` is added even if `children` are defined

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+    // No Control
+    // ----------------------------------------
+
+    if (!controlType) {
```
this if branch looks strange; can we do:
```jsx
if (!controlType) {
  return (
    <ElementType {...rest} className={classes.root}>
      {label ? labelElement : (childrenExist(children) && children)}
      {messageElement}
    </ElementType>
  )
}

bmdalex commented on src/components/Slot/Slot.tsx:
```diff
@@ -12,12 +12,18 @@ export interface ISlotProps {
 
+export const createSlotFactory = (as: any, mapValueToProps: MapValueToProps) => (
+  val,
+  options: ObjectOf<any> = {},
```
this type changed a bit, pls take a look again at #326, and also apply the fix from `factories.tsx` cause you will have a weird type error related to `IProps`

bmdalex commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,221 @@
+        typeof label === 'string'
+          ? `${label} *`
+          : typeof label === 'object' && (label as any).content
+            ? { ...label, content: `${(label as any).content} *` }
```
I don't think we should make so many presumptions on the content of the `label` shorthand; this looks like a candidate for a smaller component that will handle all this logic

mnajdova replied:
I think I over-complicated things here :\ I will just add additional text component after the label if the field is required, without changing the actual text inside the label.

bmdalex commented on types/theme.d.ts:
```diff
@@ -175,6 +175,8 @@ export interface IThemeComponentStylesInput {
   ChatItem?: IComponentPartStylesInput
   ChatMessage?: IComponentPartStylesInput
   Divider?: IComponentPartStylesInput
+  Form?: IComponentPartStylesInput
```
never have i ever update this file... I didn't even know about it; we should have a mechanism to do it for us automatically...

mnajdova replied:
Yep, this file was updated in some of the recent refactoring and adding typings for the styling. Agreed that there should be some mechanism for this.

bmdalex commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,132 @@
+    _.invoke(this.props, 'onSubmit', e, this.props, ...args)
+  }
+
+  renderFields = () => {
```
pls make private

bmdalex replied:
@mnajdova not addressed

bmdalex replied:
@mnajdova please disregard; Github is not that good at showing the correct changes..it was actually addressed

bmdalex commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,132 @@
+
+  static Field = FormField
+
+  handleSubmit = (e, ...args) => {
```
pls make private

bmdalex replied:
@mnajdova not addressed

bmdalex replied:
@mnajdova please disregard; Github is not that good at showing the correct changes..it was actually addressed

bmdalex commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,132 @@
+    as: 'form',
+  }
+
+  static Field = FormField
```
let's keep consistency within the file at least when it comes to modifiers; if we explicitly use `public` let's add it to every public member

bmdalex replied:
@mnajdova not addressed

bmdalex replied:
@mnajdova please disregard; Github is not that good at showing the correct changes..it was actually addressed

bmdalex reviewed (changes requested):
added comments and suggestions

Commit:
-refactored the code
-added the required indicator as an after of the label
-added inline and required examples

Commit:
-fixed after element for the label

Commit:
-Slot component updates

Commit:
-fixing typings

Commit:
-foxes for the id

Commit:
-foxes for the id

Commit:
-id, name, required and type are advertised from the FormField and goes inside the input element
-all rest props are going to he outer div element

Commit:
-removed unnecessary as tag

mituron committed:
changing id, adding accessibility description

mituron committed:
Merge branch 'feat/form-base' of https://github.com/stardust-ui/react into feat/form-base

mituron committed:
adding IDs which get disappear in last push

kuzhelov commented on docs/src/examples/components/Form/Types/FormExample.shorthand.tsx:
```diff
@@ -0,0 +1,40 @@
+    onSubmit={() => {
+      alert('Form submitted')
+    }}
+    fields={fields}
```
nit: when this example is rendered, it seems that there is no any gap provided between checkbox and text field:

![image](https://user-images.githubusercontent.com/9024564/47159674-acec0b00-d2ee-11e8-919a-5d6c05346e94.png)

Could we address this problem in styles?

mnajdova replied:
Will address this of course.

mnajdova replied:
Fixed for the checkbox and radio type inputs

kuzhelov replied:
Marija, this the rendering result that I had when tried the examples: 

![image](https://user-images.githubusercontent.com/9024564/47183515-86e35c80-d328-11e8-831c-5ef3ca412797.png)

Please, ping me if you'll need help to address that - probably this is result of recent changes made related to `controlType`, will be more than happy to help

mnajdova replied:
This is regression from the latest changes of the Input component. After merging the latest master, using
```jsx
<Slot as={Input} />
```
renders the html:
```html
<div class="ui-slot ui-input ui-slot">div</div>
```
It works as expected with the RadioGroup, input, Button etc. as well as with the previous implementation of the Input component. Will need to investigate this with Alex tomorrow.

kuzhelov commented on docs/src/examples/components/Form/Types/FormExample.tsx:
```diff
@@ -0,0 +1,34 @@
+  >
+    <Form.Field
+      label="First name"
+      controlType={Input}
```
this seems to have `as` semantics - maybe we should name this prop like `controlAs`?

kuzhelov replied:
also, thinking that for the first stages maybe it would be fine to eliminate this prop altogether - and just use `control` slot for that:

```jsx
 <Form.Field
      label="Last name"
      name="lastName"
      id="last-name"
      required={true}
      control={{ as: 'input' }}
```

A benefit of this approach would be that we will eliminate any possibilities for collisions of different prop values come from different places. What do you think?

mnajdova replied:
I added this as a specific property because we need it for actually creating the slot used here:
```jsx
    const factoryMethod = createSlotFactory(controlType, name => ({ name }))
    const controlElement = factoryMethod(control || {}, {
      defaultProps: { required, id, name, type, styles: styles.control },
      render: renderControl,
    })
```
We are using the Slot as this can be some HTML element, or component. If we use this form the 'as' property, then it is becoming kind a tricky because this can be object, component, some primitive type etc. What do you think? I am okay with renaming it, but yeah, let's decide whether it should exists as a prop first.

mnajdova replied:
We might right away use the Slot.create and propagate the control shorthand there, but this was the proposed way for using the Slot create method. I am open for other suggestions.

kuzhelov commented on docs/src/examples/components/Form/Types/FormExample.tsx:
```diff
@@ -0,0 +1,34 @@
+      type="checkbox"
+      id="conditions"
+    />
+    <Form.Field controlType={Button} control={{ content: 'Submit' }} />
```
proceeding with the thought mentioned above - maybe this could be considered as an alternative for the first stages?
```jsx
<Form.Field control={{ as: Button, content: 'Submit' }} />
```

mnajdova replied:
Please take a look into the previous comment: https://github.com/stardust-ui/react/pull/353#discussion_r226324506

Commit:
-fixing the styling for the input checkbox and radios

kuzhelov commented on docs/src/examples/components/Form/Types/index.tsx:
```diff
@@ -0,0 +1,20 @@
+    />
+    <ComponentExample
+      title="Inline"
+      description="The controls inside a form can appear next to the label instead of below it."
```
nit: could be simplified as `The form controls ...`

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+import FormField from './FormField'
+
+export interface IFormProps {
+  action?: string
```
nit: just for the sake of consistency with other components, lets arrange these alphabetically

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+     * A custom render iterator for rendering each of the Form fields.
+     * The default component, props, and children are available for each field.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
```
just to be sure, what are the scenarios where these JS Doc attributes (like @param) will augment IntelliSense experience? Maybe it is safe for us to just avoid them, _if_ there is no much value in them?

mnajdova replied:
They are differently displayed in the props table in the docs 
![image](https://user-images.githubusercontent.com/4512430/47219603-922d9b00-d3af-11e8-8749-bf36e0a209a7.png)

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+
+    // Heads up! Third party libs can pass own data as first argument, we need to check that it has preventDefault()
+    // method.
+    if (typeof action !== 'string') _.invoke(e, 'preventDefault')
```
could you, please, elaborate on the case we are trying to address here? How it is read by me is:
- invoke `onSubmit` props callback for the proper set of arguments being provided when form is submitted
- but if `action` is not a string, call `preventDefault` on the event object

Frankly, a bit confused with the second condition. What we are trying to handle with that? And, suppose that I am passing some garbish but `string` value as action, like `aaa` - shouldn't we call `preventDefault` in that case as well?

mnajdova replied:
This is taken from Semantic UI. The explanation for this is in the comment above.

kuzhelov replied:
yes, I've got it - but still cannot get the idea from the explanation, could you, please, help me with that? Specifically,  what am confused about is the following:

> a bit confused with the second condition. What we are trying to handle with that? And, suppose that I am passing some garbish but string value as action, like aaa - shouldn't we call preventDefault in that case as well?

mnajdova replied:
The onSubmit should be invoked only if the action is set to valid value. In other cases, the users should use onClick on some button inside the Form if they want to submit the form.

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+    // Heads up! Third party libs can pass own data as first argument, we need to check that it has preventDefault()
+    // method.
+    if (typeof action !== 'string') _.invoke(e, 'preventDefault')
+    _.invoke(this.props, 'onSubmit', e, this.props, ...args)
```
could you, please, suggest the case where these `...args` will be utilized?

mnajdova replied:
If there are some additional arguments sent here, for example if the form is going to be used with some additional library (as for validation for example) we must propagate this arguments further. Also taken from Semantic UI, made sense to me.

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+import { createSlotFactory } from '../Slot/Slot'
+
+export interface IFormFieldProps {
+  as?: any
```
nit: just to be consistent with other components, lets arrange these alphabetically

Commit:
-fixed test in the docs examples

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+    as: customPropTypes.as,
+
+    /**
+     *  Button content for childrenApi
```
oops :)

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+    as: customPropTypes.as,
+
+    /**
+     *  Form content for childrenApi
```
lets end it with dot, for the sake of consistency

Commit:
Merge branch 'master' into feat/form-base

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+    styles,
+    rest,
+  }): React.ReactNode {
+    const {
```
just sharing my thoughts on that - this seem to be the list of items (from this quite big set) one does really care about:
```jsx
const {
  children,
      control,
      label,
      id,
      name,
      type,
      required,
      message,
}  = this.props
```
This looks much cleaner

mnajdova replied:
Yep, getting rid of the renderXXX methods would reduce this much. Also, getting to experimenting to get rid of the control type :)

mnajdova replied:
Got rid of the controlType

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+    const labelElement = Text.create(label, {
+      defaultProps: {
+        as: 'label',
+        htmlFor: id,
```
👍

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+      defaultProps: {
+        styles: styles.message,
+      },
+      render: renderMessage,
```
just to be sure (but am pretty confident that this case is thought of) - it is safe to pass `null | undefined` value here, it will be just ignored, right?

mnajdova replied:
Yes, this is the case, I was testing this without sending the render method

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+      render: renderControl,
+    })
+
+    const content = (
```
now it might seem to be a matter of preference, but with more complex elements structure it will introduce unnecessary code duplication (suppose that there are three elements, instead of two `controlElement` and `labelElement` that we have now). We might consider to remove this source for code duplication in the following way:

```jsx
<>
  {isControlElementFirst && controlElement}
  {labelElement}
  {!isControlElementFirst && controlElement}

  {messageElement}
</>
```

mnajdova replied:
Agreed looks much more cleaner.

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,207 @@
+    )
+  }
+
+  private shouldControlAppearFirst = () => {
```
the name and intent is very readable 👍

Commit:
Addressed comments in PR:
-alphabetical order of the props interfaces and prop types
-improved description of the props
-rearranged the content of the FormField

Commit:
-replaced controlType with control:as

Commit:
Merge branch 'master' into feat/form-base

# Conflicts:
#	CHANGELOG.md
#	test/specs/components/Slot/Slot-test.ts

bmdalex reviewed (approved):
looks good, thx for addressing

Commit:
-removed wrapper logic from the Input, updated tests and removed wrapper examples

bmdalex reviewed (changes requested):
blocking for now just to provide better fix for the `<Slot as={Input} />` issue

bmdalex commented on src/components/Input/Input.tsx:
```diff
@@ -150,47 +134,36 @@ class Input extends AutoControlledComponent<Extendable<IInputProps>, IInputState
-      },
-      render: renderWrapper,
-    })
+    return (
```
let's not make this huge amount of changes just for the fact that we cannot use <Slot as={Input} />
the issue was that I forgot to add `children` as handled prop to the `Input` so they are passed as undefined to the `Slot.create` method; better fix is to just add:

```tsx
/**
 *  Used to set content when using childrenApi - internal only
 *  @docSiteIgnore
 */
children: PropTypes.node,
```
to the `propTypes` object in this file and
```tsx
children?: ReactChildren
```
to the `IInputProps` interface

@mnajdova @kuzhelov

mnajdova replied:
Yep, this solved the issue. Thanks!

Commit:
-reverted changes for removing the wrapper from the input and added children prop

bmdalex reviewed (approved):
looks good, let's merge 👍

kuzhelov commented on docs/src/examples/components/Form/Types/FormExample.shorthand.tsx:
```diff
@@ -0,0 +1,40 @@
+const fields = [
+  {
+    label: 'First name',
+    control: { as: Input },
```
could we provide either `Input` or `input` as a default value - for the sake of it not being necessary for the client to always specify the underlying type?

mnajdova replied:
We can provide the control to be by default `{as: Input}` but if the user wants to override it, would have to specify the as again, otherwise we would need to handle all different types of control (object, component etc) Would this be enough?

kuzhelov commented on test/specs/components/Form/FormField-test.tsx:
```diff
@@ -0,0 +1,40 @@
+  formFieldImplementsShorthandProp('message', Text)
+  formFieldImplementsShorthandProp('control', Slot)
+
+  it('renders the control provided in the controlType prop', () => {
```
now this could be renamed :) Same applies to the next test

mnajdova replied:
This will be renamed, no need for the second now that the control is a regular shorthand :)

Commit:
-added default control to be {as: Input}
-small refactoring for addressing PR comments

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,188 @@
+
+  public static defaultProps = {
+    as: 'div',
+    control: { as: Input },
```
👍

kuzhelov commented on src/components/Input/Input.tsx:
```diff
@@ -22,6 +23,7 @@ import Ref from '../Ref'
 
 export interface IInputProps {
   as?: any
+  children?: ReactChildren
```
actually, this is not correct - please, consider this comment that explains why

mnajdova replied:
This is our custom ReactChildren so it is actually the correct one.

kuzhelov commented on src/components/Form/FormField.tsx:
```diff
@@ -0,0 +1,188 @@
+
+export interface IFormFieldProps {
+  as?: any
+  children?: ReactChildren
```
this is not correct type - here is the explanation of that: https://github.com/stardust-ui/react/pull/328#discussion_r223225907

mnajdova replied:
Same as above

kuzhelov commented on src/components/Form/Form.tsx:
```diff
@@ -0,0 +1,128 @@
+export interface IFormProps {
+  action?: string
+  as?: any
+  children?: ReactChildren
```
lets fix this type (https://github.com/stardust-ui/react/pull/328#discussion_r223225907)

mnajdova replied:
Same as above

kuzhelov commented on src/index.ts:
```diff
@@ -27,6 +27,8 @@ export { default as Status } from './components/Status'
 export { default as TabBehavior } from './lib/accessibility/Behaviors/Tab/tabBehavior'
 export { default as TabListBehavior } from './lib/accessibility/Behaviors/Tab/tabListBehavior'
 export { default as Text } from './components/Text'
+export { default as Form } from './components/Form'
```
nit: those two could be grouped together `export { default as Form, FormField } from './components/Form'`

kuzhelov reviewed (approved):
couple of minor comments - please, feel free to reason whether they should be addressed

Commit:
-improved export for the Form components
